### PR TITLE
Handle trailing characters after user

### PR DESCRIPTION
**Pull request by trashcan**
This is to resolve https://github.com/mattermost/platform/issues/1130. This isn't quite complete, there is one bug left, "@myownusername..." highlights the periods for some reason. Still digging into why that is but I wanted to share what I have. I will take another stab at it tomorrow.

### CODE REVIEW

**Comment by trashcan** [comment]
Also I'm sure I did a few things wrong here, any criticism is welcome.

File: web/react/utils/text_formatting.jsx [hunk 1]
```diff
@@ -123,27 +124,60 @@ function autolinkUrls(text, tokens) {
 }
 
 function autolinkAtMentions(text, tokens) {
-    let output = text;
+    // Return true if provided character is punctuation
+    function isPunctuation(character) {
+        const re = /[\u2000-\u206F\u2E00-\u2E7F\\'!"#$%&()*+,\-.\/:;<=>?@\[\]^_`{|}~]/g;
```
Comment (bbqsrc) [inline web/react/utils/text_formatting.jsx]: `.` should be `\.` otherwise it'll match everything, no?

File: web/react/utils/text_formatting.jsx [hunk 2]
```diff
@@ -123,27 +124,60 @@ function autolinkUrls(text, tokens) {
...
+            }
+        }
+
         return fullMatch;
     }
 
-    output = output.replace(/(^|\s)(@([a-z0-9.\-_]*[a-z0-9]))/gi, replaceAtMentionWithToken);
-
+    let output = text;
+    output = output.replace(/(^|\s)(@([a-z0-9.\-_]*))/gi, replaceAtMentionWithToken);
```
Comment (bbqsrc) [inline web/react/utils/text_formatting.jsx]: Same issue as above: `.` should probably be `\.`.

File: web/react/utils/text_formatting.jsx (hunk 1 above)
Comment (trashcan) [inline web/react/utils/text_formatting.jsx]: I'm not a regex guru but I believe dots in [ ] brackets do not need to be escaped. Check out https://jsfiddle.net/21b1su24/

Comment (bbqsrc) [inline web/react/utils/text_formatting.jsx]: Heh, fair enough. How bizarre. And it seemed to fit the issue you were having too! Drats.

**Comment by it33** [comment]
Thanks @trashcan!

I think text processing in Mattermost has gotten pretty sophisticated and we need to update our tests to watch corner cases.

Towards this goal, I'm opening a new ticket for help updating the "/loadtest post fuzz": https://mattermost.atlassian.net/browse/PLT-834

We should have had something like this a while ago, and since you're probably doing some of this anyway in developing this PR, I wondered if you might be interested in taking a shot at it? https://mattermost.atlassian.net/browse/PLT-834

It'd be really helpful to everyone working on text process and it'd set a good standard and precedent. 

Also, I think being able to see all the different configurations might help you narrow down the issue you're facing--and please feel free to add more tests so you can lay out the corner cases in your fix, so that once it's done future changes don't break what you've solved?

**Comment by trashcan** [comment]
@it33, very interested thanks for doing the legwork and making the jira issue. 

I found the cause of the bug I mentioned - highlighting "..." in "@myownusername.." It turns out this is unrelated and caused by highlightCurrentMentions which I did not modify. It probably has the same original problem as autolinkAtMentions. I will work on fixing that one too.

**Comment by trashcan** [comment]
I've fixed that issue and pushed the change. I think it would be prudent to get the tests up to snuff before merging this, unless you want to.

**Comment by it33** [comment]
Thanks @trashcan!

Sorry realized we haven't documented how `/loadtest` works, I've added instructions to the ticket and here they are: 

How to use /loadtest posts fuzz:
1. Go to System Console > Enable Testing -> true -> Save
2. Go back to any channel and type `/loadtest posts fuzz` and it should run the test,

Once enabled, when you type `/loadtest` the usage should show up in auto-complete. It's off by default because there are tests you want to be careful running, because they'll create a bunch of random channels

I think some of the code from /loadtest running slash commands is "pre-historic", i.e. from a time when we were building a games portal in Django, and we want to replace it, 

At any rate, once the tests are in, you'll be able to see which issues are pre-existing and which are caused by your changes. 

The problem with multiple people working on the text process is that I think there's some change regressions are introduced with every change, and it's hard to pick that up until we get the tests in. 

Huge thanks for taking this on, I wish our documentation here were more mature. The fact that you're working on it is helping us get there.

**Comment by crspeller** [comment]
+1

**Comment by hmhealey** [comment]
Looks good. Definitely agree that we should look into adding some automated testing of stuff like this. Full UI testing of everything like we've considered in the past is definitely unnecessary, but more focused unit testing on stuff like this would be a good idea.
